---
title: "SortaDate Analysis of Acanthomorph UCEs"
author: "David Černý"
date: "3/21/2017"
output: html_document
---

# Individual UCEs

The input for a SortaDate analysis should consist of a directory of gene alignments and corresponding gene trees. Therefore, it was first necessary to estimate the individual gene trees. This was done on the local machine using RAxML as implemented in the ETE Toolkit, with a custom shell script written to repeat the following command for all the 230 loci:

```
ete3 build -w standard_raxml -n /Users/David/Grive/Alfaro_Lab/Acanthomorpha/sate-gblocks-clean-min-114-taxa/FASTA/uce-3.fasta -o /Users/David/Grive/Alfaro_Lab/Acanthomorpha/sate-gblocks-clean-min-114-taxa/uce-3-tree
```

The trees were then copied from the automatically created subdirectories into the main directory containing the alignments, and renamed so that their names matched those of the loci. SortaDate was then run on the directory in the following 4 steps:

### Phase 1: get_var_length

In the first step, the root-to-tip variance is calculated for each of the gene trees:

```
python src/get_var_length.py /Users/David/Grive/Alfaro_Lab/Acanthomorpha/sate-gblocks-clean-min-114-taxa/FASTA/ --flend .tre --outf var-uces --outg alepisaurus_ferox,ceratoscopelus_warmingii
```

Note that these two outgroups were not detected in all of the gene trees. (Showing that not all UCEs had been collected from all the taxa?) It was not indicated which trees did not contain the outgroups. Nevertheless, a `var-uces` file was still successfully produced, and contained data for all the 230 loci.

### Phase 2: get_bp_genetrees

In the second step, bipartition support values are obtained by comparing the individual gene trees with a user-supplied "master tree". This serves as a measure of topological congruence between the gene trees and the master tree:

```
python src/get_bp_genetrees.py /Users/David/Grive/Alfaro_Lab/Acanthomorpha/sate-gblocks-clean-min-114-taxa/FASTA/ /Users/David/Grive/Alfaro_Lab/Acanthomorpha/12_no_outgroups_scheme_3.tre --flend .tre --outf bp-uces
```

Note that unlike many (most?) of the gene trees, the master tree did not contain the two outgroups. This does not seem to pose a problem, since the effect on the bipartition support is the same across all trees.

### Phase 3: combine_results

In the third step, both types of data (bipartition support and root-to-tip variance) are combined to serve as the input for the last step of the analysis:

```
python src/combine_results.py var-uces bp-uces --outf comb-uces
```

```{r comment='', echo = FALSE, message=F, warning=F}
cat(readLines("comb-uces"), sep = '\n')
```

### Phase 4: get_good_genes

In order of descending priority: bipartition support, root-to-tip variance, tree length

```
python src/get_good_genes.py comb-uces --max 3 --order 3,1,2 --outf gg-uces-312
```

```{r comment='', echo = FALSE, message=F, warning=F}
cat(readLines("gg-uces-312"), sep = '\n')
```

In order of descending priority: bipartition support, tree length, root-to-tip variance

```
python src/get_good_genes.py comb-uces --max 3 --order 3,2,1 --outf gg-uces-321
```

```{r comment='', echo = FALSE, message=F, warning=F}
cat(readLines("gg-uces-321"), sep = '\n')
```

In order of descending priority: tree length, root-to-tip variance, bipartition support

```
python src/get_good_genes.py comb-uces --max 3 --order 2,1,3 --outf gg-uces-213
```

```{r comment='', echo = FALSE, message=F, warning=F}
cat(readLines("gg-uces-213"), sep = '\n')
```

In order of descending priority: tree length, bipartition support, root-to-tip variance

```
python src/get_good_genes.py comb-uces --max 3 --order 2,3,1 --outf gg-uces-231
```

```{r comment='', echo = FALSE, message=F, warning=F}
cat(readLines("gg-uces-231"), sep = '\n')
```

In order of descending priority: root-to-tip variance, tree length, bipartition support

```
python src/get_good_genes.py comb-uces --max 3 --order 1,2,3 --outf gg-uces-123
```

```{r comment='', echo = FALSE, message=F, warning=F}
cat(readLines("gg-uces-123"), sep = '\n')
```

In order of descending priority: root-to-tip variance, bipartition support, tree length

```
python src/get_good_genes.py comb-uces --max 3 --order 1,3,2 --outf gg-uces-132
```

```{r comment='', echo = FALSE, message=F, warning=F}
cat(readLines("gg-uces-132"), sep = '\n')
```

# k-means partitions

Note that the largest, slowest-evolving partition ("ccf55a6ee6d62f840a124bcc0c98ecf5"; 132 kb) was excluded from the first round of analyses for computational reasons. Attempts to analyze it in RAxML after the remaining 31 analyses finished up were unsuccessful.

### Phase 1: get_var_length

```
python src/get_var_length.py /Users/David/Grive/Alfaro_Lab/Acanthomorpha/k-means_partitions/FASTA/ --flend .tre --outf var-kmeans --outg alepisaurus_ferox,ceratoscopelus_warmingii
```

In this case, no problems with missing taxa were encountered while analyzing the partition trees.

### Phase 2: get_bp_genetrees

```
python src/get_bp_genetrees.py /Users/David/Grive/Alfaro_Lab/Acanthomorpha/k-means_partitions/FASTA/ /Users/David/Grive/Alfaro_Lab/Acanthomorpha/12_no_outgroups_scheme_3.tre --flend .tre --outf bp-kmeans
```

### Phase 3: combine_results

```
python src/combine_results.py var-kmeans bp-kmeans --outf comb-kmeans
```

```{r comment='', echo = FALSE, message=F, warning=F}
cat(readLines("comb-kmeans"), sep = '\n')
```

### Phase 4: get_good_genes

In order of descending priority: bipartition support, root-to-tip variance, tree length

```
python src/get_good_genes.py comb-kmeans --max 3 --order 3,1,2 --outf gg-kmeans-312
```

```{r comment='', echo = FALSE, message=F, warning=F}
cat(readLines("gg-kmeans-312"), sep = '\n')
```

In order of descending priority: bipartition support, tree length, root-to-tip variance

```
python src/get_good_genes.py comb-kmeans --max 3 --order 3,2,1 --outf gg-kmeans-321
```

```{r comment='', echo = FALSE, message=F, warning=F}
cat(readLines("gg-kmeans-321"), sep = '\n')
```

In order of descending priority: tree length, root-to-tip variance, bipartition support

```
python src/get_good_genes.py comb-kmeans --max 3 --order 2,1,3 --outf gg-kmeans-213
```

```{r comment='', echo = FALSE, message=F, warning=F}
cat(readLines("gg-kmeans-213"), sep = '\n')
```

In order of descending priority: tree length, bipartition support, root-to-tip variance

```
python src/get_good_genes.py comb-kmeans --max 3 --order 2,3,1 --outf gg-kmeans-231
```

```{r comment='', echo = FALSE, message=F, warning=F}
cat(readLines("gg-kmeans-231"), sep = '\n')
```

In order of descending priority: root-to-tip variance, tree length, bipartition support 

```
python src/get_good_genes.py comb-kmeans --max 3 --order 1,2,3 --outf gg-kmeans-123
```

```{r comment='', echo = FALSE, message=F, warning=F}
cat(readLines("gg-kmeans-123"), sep = '\n')
```

In order of descending priority: root-to-tip variance, bipartition support, tree length

```
python src/get_good_genes.py comb-kmeans --max 3 --order 1,3,2 --outf gg-kmeans-132
```

```{r comment='', echo = FALSE, message=F, warning=F}
cat(readLines("gg-kmeans-132"), sep = '\n')
```

It can be seen from the results (both for the individual UCEs and for the k-means partitions) that the list of clock-like genes is insensitive to the relative priority of the last two criteria, and only partially sensitive to the choice of the main criterion. In particular, tree length and root-to-tip variance tend to agree on the same set of "good genes", while prioritizing topological agreement tends to produce a completely different, non-overlapping list of loci. Note that the default setting recommended by the authors is one in which the highest priority is given to topological congruence.

# 50-bp chunks

A bash script was written to automate the following actions:

1. Create a new directory for each locus
2. Split the alignments for each locus into individual sequences, and these into 50-bp chunks using a Python script obtained from http://www.reddit.com/r/bioinformatics/comments/1u8yc7/looking_for_a_script_that_will_split_dna/ceg8rav/?st=j0tbjfco&sh=1e300055
3. For each locus, join the individual sequences chunk-wise (i.e., make a single fasta file for all taxa and sites 0 to 50, another one for all taxa and sites 51 to 100, etc.) using a custom R script
4. Add a locus-indicating prefix to all the chunks of a given UCE
5. Copy the resulting fasta files into a single directory

An additional R script was used to change the taxon names in the FASTA files so that they agree with those used in the base tree.

All the chunks were then analyzed using RAxML. In some cases, the analysis could not be performed; this might be related to the fact that some of the chunks appeared to consist exclusively of invariant sites, or that some of the taxa were represented entirely by missing data. Given that even those trees that have been successfully inferred have no bipartitions in common with the base tree, other problems are likely to have taken place in the tree reconstruction step as well.

### Phase 1: get_var_length

```
python src/get_var_length.py /Users/David/Grive/Alfaro_Lab/Acanthomorpha/sate-gblocks-clean-min-114-taxa/Chunks/ --flend .tre --outf var-chunks --outg alepisaurus_ferox,ceratoscopelus_warmingii
```

### Phase 2: get_bp_genetrees

```
python src/get_bp_genetrees.py /Users/David/Grive/Alfaro_Lab/Acanthomorpha/sate-gblocks-clean-min-114-taxa/Chunks/ /Users/David/Grive/Alfaro_Lab/Acanthomorpha/12_no_outgroups_scheme_3.tre --flend .tre --outf bp-chunks
```

### Phase 3: combine_results

```
python src/combine_results.py var-chunks bp-chunks --outf comb-chunks
```

### Phase 4: get_good_genes

In order of descending priority: bipartition support, root-to-tip variance, tree length

```
python src/get_good_genes.py comb-chunks --max 3 --order 3,1,2 --outf gg-chunks-312
```

```{r comment='', echo = FALSE, message=F, warning=F}
cat(readLines("gg-chunks-312"), sep = '\n')
```

In order of descending priority: bipartition support, tree length, root-to-tip variance

```
python src/get_good_genes.py comb-chunks --max 3 --order 3,2,1 --outf gg-chunks-321
```

```{r comment='', echo = FALSE, message=F, warning=F}
cat(readLines("gg-chunks-321"), sep = '\n')
```

In order of descending priority: tree length, root-to-tip variance, bipartition support

```
python src/get_good_genes.py comb-chunks --max 3 --order 2,1,3 --outf gg-chunks-213
```

```{r comment='', echo = FALSE, message=F, warning=F}
cat(readLines("gg-chunks-213"), sep = '\n')
```

In order of descending priority: tree length, bipartition support, root-to-tip variance

```
python src/get_good_genes.py comb-chunks --max 3 --order 2,3,1 --outf gg-chunks-231
```

```{r comment='', echo = FALSE, message=F, warning=F}
cat(readLines("gg-chunks-231"), sep = '\n')
```

In order of descending priority: root-to-tip variance, tree length, bipartition support 

```
python src/get_good_genes.py comb-chunks --max 3 --order 1,2,3 --outf gg-chunks-123
```

```{r comment='', echo = FALSE, message=F, warning=F}
cat(readLines("gg-chunks-123"), sep = '\n')
```

In order of descending priority: root-to-tip variance, bipartition support, tree length

```
python src/get_good_genes.py comb-chunks --max 3 --order 1,3,2 --outf gg-chunks-132
```

```{r comment='', echo = FALSE, message=F, warning=F}
cat(readLines("gg-chunks-132"), sep = '\n')
```

## 50-bp chunks: SH-like of 75% or above

1. Copy all the tree files to a new directory.
2. Collapse all the nodes with SH-like support values of less than 75% using a Python script.
3. Copy the fasta alignments to the same directory.
    
### Phase 1: get_var_length

```
python src/get_var_length.py /Users/David/Grive/Alfaro_Lab/Acanthomorpha/sate-gblocks-clean-min-114-taxa/Chunks_75/ --flend .tre --outf var-chunks75 --outg alepisaurus_ferox,ceratoscopelus_warmingii
```

### Phase 2: get_bp_genetrees

```
python src/get_bp_genetrees.py /Users/David/Grive/Alfaro_Lab/Acanthomorpha/sate-gblocks-clean-min-114-taxa/Chunks_75/ /Users/David/Grive/Alfaro_Lab/Acanthomorpha/12_no_outgroups_scheme_3.tre --flend .tre --outf bp-chunks75
```

### Phase 3: combine_results

```
python src/combine_results.py var-chunks75 bp-chunks75 --outf comb-chunks75
```

(Use an R script to delete the lines containing NAs.)

### Phase 4: get_good_genes

In order of descending priority: bipartition support, root-to-tip variance, tree length

```
python src/get_good_genes.py comb-chunks75-filtered --max 3 --order 3,1,2 --outf gg-chunks75-312
```

```{r comment='', echo = FALSE, message=F, warning=F}
cat(readLines("gg-chunks75-312"), sep = '\n')
```

In order of descending priority: bipartition support, tree length, root-to-tip variance

```
python src/get_good_genes.py comb-chunks75-filtered --max 3 --order 3,2,1 --outf gg-chunks75-321
```

```{r comment='', echo = FALSE, message=F, warning=F}
cat(readLines("gg-chunks75-321"), sep = '\n')
```

In order of descending priority: tree length, root-to-tip variance, bipartition support

```
python src/get_good_genes.py comb-chunks75-filtered --max 3 --order 2,1,3 --outf gg-chunks75-213
```

```{r comment='', echo = FALSE, message=F, warning=F}
cat(readLines("gg-chunks75-213"), sep = '\n')
```

In order of descending priority: tree length, bipartition support, root-to-tip variance

```
python src/get_good_genes.py comb-chunks75-filtered --max 3 --order 2,3,1 --outf gg-chunks75-231
```

```{r comment='', echo = FALSE, message=F, warning=F}
cat(readLines("gg-chunks75-231"), sep = '\n')
```

In order of descending priority: root-to-tip variance, tree length, bipartition support 

```
python src/get_good_genes.py comb-chunks75-filtered --max 3 --order 1,2,3 --outf gg-chunks75-123
```

```{r comment='', echo = FALSE, message=F, warning=F}
cat(readLines("gg-chunks75-123"), sep = '\n')
```

In order of descending priority: root-to-tip variance, bipartition support, tree length

```
python src/get_good_genes.py comb-chunks75-filtered --max 3 --order 1,3,2 --outf gg-chunks75-132
```

```{r comment='', echo = FALSE, message=F, warning=F}
cat(readLines("gg-chunks75-132"), sep = '\n')
```

## 50-bp chunks: SH-like of 90% or above

(The first three steps were identical to those described above.)
    
### Phase 1: get_var_length

```
python src/get_var_length.py /Users/David/Grive/Alfaro_Lab/Acanthomorpha/sate-gblocks-clean-min-114-taxa/Chunks_90/ --flend .tre --outf var-chunks90 --outg alepisaurus_ferox,ceratoscopelus_warmingii
```

### Phase 2: get_bp_genetrees

```
python src/get_bp_genetrees.py /Users/David/Grive/Alfaro_Lab/Acanthomorpha/sate-gblocks-clean-min-114-taxa/Chunks_90/ /Users/David/Grive/Alfaro_Lab/Acanthomorpha/12_no_outgroups_scheme_3.tre --flend .tre --outf bp-chunks90
```

### Phase 3: combine_results

```
python src/combine_results.py var-chunks90 bp-chunks90 --outf comb-chunks90
```

(Delete the lines with NAs)

### Phase 4: get_good_genes

In order of descending priority: bipartition support, root-to-tip variance, tree length

```
python src/get_good_genes.py comb-chunks90-filtered --max 3 --order 3,1,2 --outf gg-chunks90-312
```

```{r comment='', echo = FALSE, message=F, warning=F}
cat(readLines("gg-chunks90-312"), sep = '\n')
```

In order of descending priority: bipartition support, tree length, root-to-tip variance

```
python src/get_good_genes.py comb-chunks90-filtered --max 3 --order 3,2,1 --outf gg-chunks90-321
```

```{r comment='', echo = FALSE, message=F, warning=F}
cat(readLines("gg-chunks90-321"), sep = '\n')
```

In order of descending priority: tree length, root-to-tip variance, bipartition support

```
python src/get_good_genes.py comb-chunks90-filtered --max 3 --order 2,1,3 --outf gg-chunks90-213
```

```{r comment='', echo = FALSE, message=F, warning=F}
cat(readLines("gg-chunks90-213"), sep = '\n')
```

In order of descending priority: tree length, bipartition support, root-to-tip variance

```
python src/get_good_genes.py comb-chunks90-filtered --max 3 --order 2,3,1 --outf gg-chunks90-231
```

```{r comment='', echo = FALSE, message=F, warning=F}
cat(readLines("gg-chunks90-231"), sep = '\n')
```

In order of descending priority: root-to-tip variance, tree length, bipartition support 

```
python src/get_good_genes.py comb-chunks90-filtered --max 3 --order 1,2,3 --outf gg-chunks90-123
```

```{r comment='', echo = FALSE, message=F, warning=F}
cat(readLines("gg-chunks90-123"), sep = '\n')
```

In order of descending priority: root-to-tip variance, bipartition support, tree length

```
python src/get_good_genes.py comb-chunks90-filtered --max 3 --order 1,3,2 --outf gg-chunks90-132
```

```{r comment='', echo = FALSE, message=F, warning=F}
cat(readLines("gg-chunks90-132"), sep = '\n')
```